### ISSUE TITLE
Add segment length property to Coordinates1d

### ISSUE THREAD

**Issue by jmilloy** [creare-com/podpac#120]
**Description**
Non-monotonic coordinates cannot use segment ctype. In particular, area_bounds and intersection with outer=True do not work.

**Describe the solution you'd like**
Add a segment_length property, which can be used to implement area_bounds and outer intersections.

**Sample Code**

```
>>> c = ArrayCoordinates1d([10, 4, 6], ctype='center', segment_length=2)
>>> c.area_bounds
[3, 11]
```

**Comment by jmilloy** [comment]
This issue was discussed in #109 and will replace the workaround in 7811d6c.

**Comment by mpu-creare** [comment]
Incidentally, I suppose the segment_length should be allowed to be an array the same size as the coordinates. That way you can set the size of each segment in a non-monotonic system.

**Comment by jmilloy** [comment]
Sure thing. What about UniformCoordinates? Should segment_length (value or array) be supported in the same way, or is the segment length always going to be the step?

**Comment by mpu-creare** [comment]
Mmm.. that's a good point. Although, it would be an odd case to have different segment lengths for uniform coordinates.... but it could happen. 

I think I'd rather not try to support this functionality because I doubt anyone will ever use it. I think calculating segment_length (i.e. == step) in UniformCoordinates is fine. If we ever do have a use-case, I don't think the refactor would be horrendous. 

Waffling here... if it helps to make the interface more uniform, I also have no problem supporting this functionality in UniformCoordinates. So I leave it up to you what you want to do.

**Comment by jmilloy** [comment]
@mpu-creare I started to look at this again this morning. I don't think that segment lengths are required for non-monotonic outer intersection. I think it is only the area_bounds that would use segment lengths, but area_bounds already uses extents if they were provided, so the segment lengths wouldn't be adding any functionality. I'm going to implement the non-monotonic outer intersection and submit a PR. I don't want to spend time on a general segment lengths feature if it is not yet needed, but I want to check with you because there is maybe something that I am missing.

**Comment by mpu-creare** [comment]
* In the monotonic case, we actually know the bounds, and they should come from the neighboring elements:
``` python
In [19]: c = ArrayCoordinates1d([1, 2, 5, 7], extents=[0, 8])

In [20]: c[2].coordinates
Out[20]: array([5.])

In [21]: c[2].extents
Out[21]: array([3., 6.])
```
So this suggests maintaining the segment lengths

* Same goes for the strided selection:
```python
>>> c = ArrayCoordinates1d([0, 1, 2, 3, 4])
>>> c.area_bounds
array([-0.5,  4.5])
>>> c[1::2].coordinates
array([1., 3.])
>>> c[1::2].area_bounds
array([0.5,  3.5])
```

The strided selection is not great, because you'll really have gappy data at this point.
I.e. starting from: 

``` 
-------
You'll have the following after strided selection
- - - -
```

i.e. gappy coverage. This is the correct behavior because the coordinates describe the coverage of the data. Unless you do an averaging operation (using a node for example), the coordinates will refer to a cell, which doesn't change size when you sub-select. 

So I think the answer is clearly **MAINTAIN** segment lengths. 

In the non-monotonic case when you have to provide segment lengths (because we have no way of inferring anything), we can just maintain the segment lengths after selection and then the behavior will be consistent.

**Comment by jmilloy** [comment]
Great, that's what makes logical sense to me, as well.

The most straightforward way to implement this is a `segment_lengths` trait with a @tl.default defined that infers the lengths when possible, but that will roughly double the size of array coordinates. Is that okay?

Do you want to require segment_lengths when creating singleton coordinates, as well, or is there some default that makes sense? For datetime, the segment length default could be the resolution, e.g. '2018-01-01' would default to '1D' segments and '2018-01-01 12:00' would default to '1m' segmentsc, but it's harder to see what that should be for a numerical coordinate because we're using floats (instead of decimals).

**Comment by mpu-creare** [comment]
can we make `segments_lengths` either a float or an array? In that case it won't double the size? 

Doubling the size is fine. If you can't handle the size of double the array you have bigger problems. 

I think we should require `segment_length` when creating coordinates of type `not point`. I don't see another way around that. The default for singletons should probably be `point`... I bet that will lead to some issues with the SMAP node, but it will have to be fixed. I don't think we should infer the segment length of a time coordinate -- even though I've been doing that in some places. The conversion from `datetime` to `np.datetime64` is shaky enough that we won't get a good, consistent resolution and I bet it will catch users off-guard.

**Comment by jmilloy** [comment]
I'll try. I'll probably implement arrays first and do the optimization separately.

I learned yesterday that traitlets validators are *not* called when you set a trait to None. That was making this a bit of a pain to implement well. I just had a new thought, though, and it *does* call the validator when you set a trait to None if the trait is `tl.Any`, so I can use that instead of a Union.

I remember I had trouble setting the default ctype to be 'point' for singletons, but I will revisit that and see if I can make it work.

**Comment by jmilloy** [comment]
Implemented in #163
